bin/varnishd/cache/cache_esi_deliver.c

```diff
@@ -311,6 +311,9 @@ ved_vdp_esi_bytes(struct req *req, enum vdp_action act, void **priv,
 	struct ecx *ecx;
 	int retval = 0;
 
+	if (act == VDP_END)
```

Comment:
Looking at how I did this, I have the following snippet (I had it named VDP_LAST, but that is the same meaning as VDP_END):

```
	if (act == VDP_LAST) {
		/* It's tricky to know when to signal VDP_LAST down the
		 * VDP stack. Run with VDP_FLUSH, and signal an empty
		 * VDP_LAST afterwards. */
		retval = ved_vdp_esi_bytes(req, VDP_FLUSH, priv, ptr, len);
		if (retval == VDP_OK)
			retval = VDP_bytes(req, VDP_LAST, NULL, 0);
		return (retval);
	}

```

Point being that the VDPs of the suburequests should be able to see a VDP_END event if they so require. But of course this should not be propagated to the parent request during an ESI subrequest handling, and that will then be filtered away again in ved_bytes() (I see this filter is already present). So I think the snippet given here needs to be added.

Reply:
Right.
This only concerns state 3 (literal bytes), because the other `VDP_Bytes` calls already have explicit actions, but I agree that my patch did not get that case right if the literal bytes came last.
Also, by your requirement that the filters on all ESI levels need to get to see 'VDP_END`, I got case 2 wrong in de6288cbe4ac102ebd3ca66cddb64bc12069ee49.
Yet I would want to avoid the zero-length `VDP_END` by your own argument. So I have pushed a different patch.

Reply:
I went back and forth a bit, but now addressed the relevant issues both in master and in this PR.